test/jasmine/assets/custom_assertions.js

```diff
@@ -216,3 +216,29 @@ exports.assertElemInside = function(elem, container, msg) {
       contBB.top < elemBB.top &&
       contBB.bottom > elemBB.bottom).toBe(true, msg);
 };
+
+/*
+ * quick plot area dimension check: test width and/or height of the inner
+ * plot area (single subplot) to verify that the margins are as expected
+ * opts can have keys (all optional):
+ *   width (exact width match)
+ *   height (exact height match)
+ *   widthLessThan (width must be less than this)
+ *   heightLessThan (height must be less than this)
+ */
+exports.assertPlotSize = function(opts, msg) {
+    var width = opts.width;
+    var height = opts.height;
+    var widthLessThan = opts.widthLessThan;
+    var heightLessThan = opts.heightLessThan;
+
+    var actualWidth = d3.select('.ygrid').node().getBoundingClientRect().width;
+    var actualHeight = d3.select('.xgrid').node().getBoundingClientRect().height;
```

Comment:
(non-blocking) but wouldn't `d3.select('.subplot.xy > .plot').node().getBoundingClientRect` be more general (e.g. would work even on graphs with `showgrid: false`)?

Reply:
Hmm, good point, but `.plot` shrinks to the data, which may not reach the edges of the plot area (or could even go beyond with `cliponaxis: false`). Possibly `.bglayer .bg` though...

Reply:
> or could even go beyond with cliponaxis: false

Yeah, good point here. Though problem it turns out.

Reply:
switched to `.bglayer .bg` in https://github.com/plotly/plotly.js/pull/2681/commits/bb6c7201820468747976de41baa90f65a08eb4ee though this has a caveat about `margin.pad` https://github.com/plotly/plotly.js/pull/2681/commits/5d3132423a1a4a347528d80feb03ba6ad7bc356e. I still think it's an improvement: `margin.pad` isn't used very much and can be corrected for if you really want, and the bg rect is always present.